core/src/main/java/io/questdb/cairo/TxnScoreboardPoolV2.java

```diff
@@ -0,0 +1,130 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo;
+
+
+import io.questdb.std.ConcurrentHashMap;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+
+import static io.questdb.cairo.pool.AbstractMultiTenantPool.ENTRY_SIZE;
+
+public class TxnScoreboardPoolV2 implements TxnScoreboardPool {
+    private final BiFunction<CharSequence, ScoreboardPoolTenant, ScoreboardPoolTenant> getOrCreateScoreboard;
+    private final ConcurrentHashMap<ScoreboardPoolTenant> pool = new ConcurrentHashMap<>();
+
+    public TxnScoreboardPoolV2(CairoConfiguration configuration) {
+        getOrCreateScoreboard = (key, value) -> {
+            if (value == null || !value.incrementRefCount()) {
+                //noinspection resource
+                value = new ScoreboardPoolTenant(configuration.getReaderPoolMaxSegments() * ENTRY_SIZE);
+            }
+            return value;
+        };
+    }
+
+    @Override
+    public void clear() {
+        for (CharSequence tableDir : pool.keySet()) {
+            final var scoreboard = pool.get(tableDir);
+            if (scoreboard != null) {
+                // Don't use iterator.remove(), it can remove new entry
+                pool.remove(tableDir, scoreboard);
+                if (!scoreboard.tryFullClose()) {
+                    scoreboard.closePending = true;
+                    scoreboard.tryFullClose();
```

Comment:
nit: this call was missing, so I added it. Usually, when a volatile flag is set, the action should be repeated as there is a good chance that the other thread(s) won't see the new flag value.